### PR TITLE
make MIR type checker handle a number of other cases

### PR DESCRIPTION
**Pull request by nikomatsakis**
The existing type checker was primarily used to verify types, but was skipping over a number of details. For example, it was not checking that the predicates on functions were satisfied and so forth. This meant that the NLL region checker was not getting a lot of the constraints it needed. This PR closes those gaps. It also includes a bit of refactoring for the way that we store region values, encapsulating the bit matrix over into its own module and improving the data structures in use.

This is mostly work by @spastorino being ported over from nll-master.

r? @arielb1 or @pnkfelix

### CODE REVIEW

File: src/librustc_mir/transform/type_check.rs [hunk 1]
```diff
@@ -1110,21 +1151,123 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
...
+                    // The type that we see in the fcx is like
+                    // `foo::<'a, 'b>`, where `foo` is the path to a
+                    // function definition. When we extract the
+                    // signature, it comes from the `fn_sig` query,
+                    // and hence may contain unnormalized results.
+                    let fn_sig = self.normalize(&fn_sig, location);
+
+                    let ty_fn_ptr_from = tcx.mk_fn_ptr(fn_sig);
+
+                    if let Err(terr) = self.eq_types(ty_fn_ptr_from, ty, location.at_self()) {
```
Comment (arielb1) [inline src/librustc_mir/transform/type_check.rs]: I think typeck allows some sort of (e.g. higher-ranked) subtyping here, so you'll need to allow it too.

e.g.
```Rust
// this sig contains a binder
fn foo<'a>(_x: &'a ()) {}
fn bar<'b>() {
    foo as fn(&'b ()); // ... but this doesn't
}
```

File: src/test/compile-fail/mir_check_cast_reify.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+{
+    panic!()
+}
+
+fn bar<'a>(x: &'a u32) -> &'static u32 {
+    // Here, the type of `foo` is `typeof(foo::<'x>)` for some fresh variable `'x`.
+    // During NLL region analysis, this will get renumbered to `typeof(foo::<'?0>)`
+    // where `'?0` is a new region variable.
+    //
+    // (Note that if `'a` on `foo` were early-bound, the type would be
```
Comment (arielb1) [inline src/test/compile-fail/mir_check_cast_reify.rs]: ... but 'a *is* early-bound. Did you mean late-bound?

**Review by arielb1 — CHANGES_REQUESTED** [review]
see comments. Nice PR otherwise.

File: src/librustc_mir/borrow_check/nll/universal_regions.rs [hunk 3]
```diff
@@ -493,7 +493,27 @@ impl<'cx, 'gcx, 'tcx> UniversalRegionsBuilder<'cx, 'gcx, 'tcx> {
...
+            // expressions, such as the `22` in `[foo; 22]`, we can
+            // get the type `usize` here. For now, just return an
+            // empty vector of substs in this case, since there are no
+            // generics in scope in such expressions right now.
+            //
+            // Eventually I imagine we could get a wider range of
+            // types.  What is the best way to handle this? Should we
+            // be checking something other than the type of the def-id
+            // to figure out what to do (e.g. the def-key?).
+            ty::TyUint(..) => {
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply src/librustc_mir/borrow_check/nll/universal_regions.rs]: It ICEs. I have later patches that do better.

File: src/test/compile-fail/mir_check_cast_reify.rs (hunk 2 above)
Comment (nikomatsakis) [reply src/test/compile-fail/mir_check_cast_reify.rs]: I did mean late-bound -- if it were late-bound, it wouldn't be part of the type of `foo`

File: src/librustc_mir/transform/type_check.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/librustc_mir/transform/type_check.rs]: Hmm, that code type-checks. Perhaps because I haven't addressed previous comment yet.

**Comment by nikomatsakis** [comment]
OK @arielb1 pushed some comments to address those.

**Comment by nikomatsakis** [comment]
rebased, fixed a few minor things relating to changes in the underlying borrowck leading to slightly different errors, but I kept the separate commits in place

**Comment by arielb1** [comment]
Error message needs fixing:
```
[00:59:35] 
[00:59:35] ---- [compile-fail] compile-fail/nll/where_clauses_in_repeat_rvalue.rs stdout ----
[00:59:35] 	
[00:59:35] error: /checkout/src/test/compile-fail/nll/where_clauses_in_repeat_rvalue.rs:30: unexpected error: '30:17: 30:19: `x` does not live long enough [E0597]'
[00:59:35] 
[00:59:35] error: /checkout/src/test/compile-fail/nll/where_clauses_in_repeat_rvalue.rs:38: expected error not found: borrowed value does not live long enough [E0597]
[00:59:35] 
[00:59:35] error: 1 unexpected errors found, 1 expected errors not found
[00:59:35] status: exit code: 101
[00:59:35] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/compile-fail/nll/where_clauses_in_repeat_rvalue.rs" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/compile-fail" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/compile-fail/nll/where_clauses_in_repeat_rvalue.stage2-x86_64-unknown-linux-gnu" "-Crpath" "-O" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Z" "borrowck=mir" "-Z" "nll" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/compile-fail/nll/where_clauses_in_repeat_rvalue.stage2-x86_64-unknown-linux-gnu.aux" "-A" "unused"
[00:59:35] unexpected errors (from JSON output): [
[00:59:35]     Error {
[00:59:35]         line_num: 30,
[00:59:35]         kind: Some(
[00:59:35]             Error
[00:59:35]         ),
[00:59:35]         msg: "30:17: 30:19: `x` does not live long enough [E0597]"
[00:59:35]     }
[00:59:35] ]
[00:59:35] 
[00:59:35] not found errors (from test file): [
[00:59:35]     Error {
[00:59:35]         line_num: 38,
[00:59:35]         kind: Some(
[00:59:35]             Error
[00:59:35]         ),
[00:59:35]         msg: "borrowed value does not live long enough [E0597]"
[00:59:35]     }
[00:59:35] ]
[00:59:35] 
[00:59:35] thread '[compile-fail] compile-fail/nll/where_clauses_in_repeat_rvalue.rs' panicked at 'explicit panic', /checkout/src/tools/compiletest/src/runtest.rs:1163:12
[00:59:35] 
[00:59:35] 
[00:59:35] failures:
[00:59:35]     [compile-fail] compile-fail/nll/where_clauses_in_repeat_rvalue.rs
[00:59:35] 
[00:59:35] test result: [31mFAILED(B[m. 2698 passed; 1 failed; 15 ignored; 0 measured; 0 filtered out
[00:59:35] 
```

**Comment by nikomatsakis** [comment]
@arielb1 pushed an update to that error msg...let's see what travis thinks

**Pull request merged**